portfolio/src/main/java/com/google/sps/servlets/NameServlet.java

```diff
@@ -0,0 +1,106 @@
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/name")
+public class NameServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    
+    UserService userService = UserServiceFactory.getUserService();
+    if (userService.isUserLoggedIn()) {
+        String userId = userService.getCurrentUser().getUserId();
+        String firstName = getFirstName(userId);
+        String lastName = getLastName(userId);
+      
+        // need to set a name
+        if(firstName.equals("") && lastName.equals("")) {
+            response.sendRedirect("/name.html");
+        }
+        else {
+            response.sendRedirect("/contact.html");
+        }
+    } 
+    else {
+        String loginUrl = userService.createLoginURL("/name");
+        out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+    }
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/name");
+      return;
+    }
+
+    String firstName = request.getParameter("firstName");
+    String lastName = request.getParameter("lastName");
+    String id = userService.getCurrentUser().getUserId();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity entity = new Entity("UserInfo", id);
+    entity.setProperty("id", id);
+    entity.setProperty("firstName", firstName);
+    entity.setProperty("lastName", lastName);
+    datastore.put(entity);
+
+    response.sendRedirect("/contact.html");
+  }
+
+  // returns the first name of the user with id, or empty String if the user has not set a nickname.
+  public static String getFirstName(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("UserInfo").setFilter(new Query.FilterPredicate("id", Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return "";
+    }
```

Comment:
This code seems to be repeated. Maybe create a function that gets the Entity from the id? Then you can have functions that get the first or last name from it.